worlds/jakanddaxter/Rules.py

```diff
@@ -0,0 +1,222 @@
+import typing
+from BaseClasses import CollectionState
+from Options import OptionError
+from . import JakAndDaxterWorld
+from .Options import (EnableOrbsanity,
+                      GlobalOrbsanityBundleSize,
+                      PerLevelOrbsanityBundleSize,
+                      FireCanyonCellCount,
+                      MountainPassCellCount,
+                      LavaTubeCellCount,
+                      CitizenOrbTradeAmount,
+                      OracleOrbTradeAmount)
+from .locs import CellLocations as Cells
+from .Locations import location_table
+from .Levels import level_table
+
+
+def set_orb_trade_rule(world: JakAndDaxterWorld):
+    options = world.options
+    player = world.player
+
+    if options.enable_orbsanity == EnableOrbsanity.option_off:
+        world.can_trade = lambda state, required_orbs, required_previous_trade: (
+            can_trade_vanilla(state, player, world, required_orbs, required_previous_trade))
+    else:
+        world.can_trade = lambda state, required_orbs, required_previous_trade: (
+            can_trade_orbsanity(state, player, world, required_orbs, required_previous_trade))
+
+
+def recalculate_reachable_orbs(state: CollectionState, player: int, world: JakAndDaxterWorld) -> None:
+
+    if not state.prog_items[player]["Reachable Orbs Fresh"]:
+
+        # Recalculate every level, every time the cache is stale, because you don't know
+        # when a specific bundle of orbs in one level may unlock access to another.
+        for level in level_table:
+            state.prog_items[player][f"{level} Reachable Orbs".strip()] = (
+                count_reachable_orbs_level(state, world, level))
+
+        # Also recalculate the global count, still used even when Orbsanity is Off.
+        state.prog_items[player]["Reachable Orbs"] = count_reachable_orbs_global(state, world)
```

Comment:
So, um, forgive me for a potentially stupid question, but isn't the global count just the sum of the level counts? So, couldn't you have `count_reachable_orbs_level` return the count it calculated, and add that onto a running sum? 

Even if it's not, you could still sum up the level ones, and have it calculate the non-level orbs at the end and add it to the sum

That should be a pretty big performance boost if I'm not mistaken

Reply:
There are in fact no orbs in Jak and Daxter that are not part of a level. So why didn't I do it your way to begin with... I have no idea. I have a vague sensation that there must have been some reason, that I already tried, or something. But I'll give it another go and show the results if it doesn't turn out well.